docs/proposals/signal-api.md

```diff
@@ -0,0 +1,172 @@
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Signal API
+
+This document describes signal API effort in Kubernetes.
+
+## Abstract
+
+The goal is to provide a possibility to send signals to container entrypoint application. We are going to introduce
+a concept of *notification* in the container's definition inside pod's spec, which will provide a possibility of
+implementing the *notification types*. One of these types will be signals. Container will have a list of notifications
+with the user-defined name (and also type-specific intormations), then the API resource will allow to use that
+notification.
+
+In case of signal type, the information which will be provided to this notification list is a POSIX signal name.
+The signal notification will be in fact a mapping from the notification name to the actual POSIX name. In example,
+user may want to define notification called *bump*, give it a *signal* type and point it to the SIGHUP signal.
+After creating such a pod, he would be able to send it a *bump* notification, which will result.
+
+The idea of the notifications allows to create many types, but only implementing the signal type is in the scope
+of this proposal.
+
+Once me make agreement on the API part, the document will be extended with the implementation details.
+
+One thing that needs to be noticed is that restarting scenarios are not a goad of this proposal. The only goal here
+is to send single signals, no more, no less. More complicated restart scenarios (like sending SIGTERM, then waiting
+for graceful termination, then sending SIGKILL after reaching timeout) may need another resource in Kubernetes API,
+which will not expose any POSIX terminology to the user.
+
+## Motivation
+
+Many Kubernetes users and operators would like to send custom signals to the processes running in containers. That
+may be motivated both by default behaviour that signals are bringing and the custom signal handling which the
+program is doing.
+
+The more concrete use-cases are described below.
+
+## Use-Cases
+
+#### Re-reading configuration
+
+Many applications follow the convention to re-read configuration files when handling the SIGHUP signal. Ability
+to send SIGHUP signal will allow us to reconfigure pods without updating their definition or removing them. Also,
+we can consider using this API in the watch mechanism in kubelet to reconfigure a pod automatically on the ConfigMap
+change.
+
+## Analysis
+
+The system needs to be able to:
+
+1. Model correctly which signals are supported
+2. Allow to define what signals can be used for each pod (and whether they can be used at all)
+3. Provide custom names for the chosen signals in pod
```

Comment:
I'm not sure this need follows obviously from the above. Yeah, it makes the API pretty, but it also seems like this could lead to e.g. some people having `reload -> SIGHUP` and others having `reconfigure -> SIGHUP`, so the prettyness of the API will also end up with the possibility of somewhat inconsistent usage.

If I'm writing code to SIGHUP an application, I'll have to parse the podspec to figure out which signal name maps to it. Conversely, if I want to "reload" the application configuration, I'll have no way of being sure that `signal reload` will be named as such.

I'm not against the idea of named signals per-se, I just feel like the need for this choice isn't expressed clearly anywhere.